packages/promotion/src/utils/compute-actions/items.ts

```diff
@@ -0,0 +1,139 @@
+import {
+  ApplicationMethodAllocationValues,
+  PromotionTypes,
+} from "@medusajs/types"
+import {
+  ApplicationMethodAllocation,
+  ApplicationMethodTargetType,
+  MedusaError,
+} from "@medusajs/utils"
+import { areRulesValidForContext } from "../validations"
+
+export function getComputedActionsForItems(
+  promotion: PromotionTypes.PromotionDTO,
+  itemApplicationContext: PromotionTypes.ComputeActionContext[ApplicationMethodTargetType.ITEMS],
+  itemIdPromoValueMap: Map<string, number>,
+  allocationOverride?: ApplicationMethodAllocationValues
+): PromotionTypes.ComputeActions[] {
+  const applicableItems: PromotionTypes.ComputeActionContext[ApplicationMethodTargetType.ITEMS] =
+    []
+
+  if (!itemApplicationContext) {
+    throw new MedusaError(
+      MedusaError.Types.INVALID_DATA,
+      `"items" not found in context`
+    )
+  }
+
+  for (const itemContext of itemApplicationContext) {
+    const isPromotionApplicableToItem = areRulesValidForContext(
+      promotion?.application_method?.target_rules!,
+      itemContext
+    )
+
+    if (!isPromotionApplicableToItem) {
+      continue
+    }
+
+    applicableItems.push(itemContext)
+  }
+
+  return applyPromotionToItems(
+    promotion,
+    applicableItems,
+    itemIdPromoValueMap,
+    allocationOverride
+  )
+}
+
+export function applyPromotionToItems(
+  promotion: PromotionTypes.PromotionDTO,
+  items: PromotionTypes.ComputeActionContext[ApplicationMethodTargetType.ITEMS],
+  itemIdPromoValueMap: Map<string, number>,
+  allocationOverride?: ApplicationMethodAllocationValues
+): PromotionTypes.ComputeActions[] {
+  const { application_method: applicationMethod } = promotion
+  const allocation = applicationMethod?.allocation!
+  const computedActions: PromotionTypes.ComputeActions[] = []
+
+  if (
+    [allocation, allocationOverride].includes(ApplicationMethodAllocation.EACH)
+  ) {
+    for (const item of items!) {
+      const appliedPromoValue = itemIdPromoValueMap.get(item.id) || 0
+      const promotionValue = parseFloat(applicationMethod!.value!)
+      const applicableTotal =
+        item.unit_price *
+          Math.min(item.quantity, applicationMethod?.max_quantity!) -
+        appliedPromoValue
+
+      let amount = promotionValue
+
+      if (promotionValue > applicableTotal) {
+        amount = applicableTotal
+      }
+
+      if (amount <= 0) {
```

Comment:
question: does that mean that we can't apply a value that is greater than the price? otherwise we can do `Math.max(applicableTotal, 0)`

Reply:
If those comments looks valid, they can be applied down to

Reply:
That would be incorrect. Math.min is a good suggestion for getting the least value to apply, but here if the amount is 0 or less, we should skip compute actions. The promotion could have been already applied in another promotion in the same array.

Reply:
but ok for the explantion I get it

Reply:
From your suggestion on above :)